File: backend/app/schemas/trace.py
"""Agent trace schema.

Every ``/api/chat`` response includes an ``AgentTrace`` even in demo mode
(AGENTS.md §13). The trace keeps the Agent / Tool / Guard / StateEvent /
Retrieval / Memory distinction visible (AGENTS.md §5). Slice 1 emits a minimal
baseline trace; #9 will expand persistence and the Trace Panel.
"""

from __future__ import annotations

from typing import Any, Optional

from pydantic import BaseModel, Field

from app.core.constants import CompanionMode, RiskLevel, Route, TraceEntryKind


class TraceStep(BaseModel):
    """One step in the trace, tagged with its ``kind`` so the UI can label it."""

    kind: TraceEntryKind
    name: str
    summary: str = ""
    detail: dict[str, Any] = Field(default_factory=dict)


class ResearchRoleTrace(BaseModel):
    selected_roles: list[str] = Field(default_factory=list)
    primary_role: Optional[str] = None
    role_selection_mode: Optional[str] = None
    requested_role_ids: list[str] = Field(default_factory=list)
    cueing_style: Optional[str] = None


class ResearchTopicTrace(BaseModel):
    topic_id: Optional[str] = None
    topic_label: Optional[str] = None
    material_type: Optional[str] = None
    classified_topic: Optional[str] = None


class ResearchBoundaryTrace(BaseModel):
    boundary_state: str = "none"
    boundary_notes: list[str] = Field(default_factory=list)


class ResearchInteractionTrace(BaseModel):
    intent: Optional[str] = None
    role_selection_source: Optional[str] = None
    context_role_ids: list[str] = Field(default_factory=list)


class ResearchControlTrace(BaseModel):
    study_condition: Optional[str] = None
    study_session_id: Optional[str] = None
    elder_control_action: Optional[str] = None


class ResearchTraceMetadata(BaseModel):
    interaction: ResearchInteractionTrace = Field(
        default_factory=ResearchInteractionTrace
    )
    role: ResearchRoleTrace = Field(default_factory=ResearchRoleTrace)
    topic: ResearchTopicTrace = Field(default_factory=ResearchTopicTrace)
    boundary: ResearchBoundaryTrace = Field(default_factory=ResearchBoundaryTrace)
    control: ResearchControlTrace = Field(default_factory=ResearchControlTrace)


class AgentTrace(BaseModel):
    turn_id: str
    mode: CompanionMode
    route: Route
    risk_level: RiskLevel
    agents: list[TraceStep] = Field(default_factory=list)
    tools: list[TraceStep] = Field(default_factory=list)
    guards: list[TraceStep] = Field(default_factory=list)
    state_event: Optional[TraceStep] = None
    memory_used: bool = False
    retrieval_used: bool = False
    safety_critic_used: bool = False
    conversation_history_used: bool = False
    conversation_history_count: int = 0
    conversation_seed_used: bool = False
    conversation_seed_count: int = 0
    # Compact research/demo metadata for relationship-aware turns. Keep this
    # privacy-preserving: IDs/enums/role choices, not full sensitive content.
    research_metadata: dict[str, Any] = Field(default_factory=dict)
    research_trace: ResearchTraceMetadata = Field(default_factory=ResearchTraceMetadata)


class TraceRecord(BaseModel):
    """A persisted trace: the per-turn AgentTrace plus who/when (issue #9)."""

    turn_id: str
    user_id: str
    created_at: str  # ISO-8601 UTC
    trace: AgentTrace


class TraceSummary(BaseModel):
    """Compact row for the trace history list."""

    turn_id: str
    user_id: str
    created_at: str
    route: Route
    risk_level: RiskLevel
    safety_critic_used: bool = False
    retrieval_used: bool = False

File: frontend/src/types/trace.ts
// Mirrors backend app/schemas/trace.py. Keeps the Agent / Tool / Guard
// distinction visible (AGENTS.md §5, §13).

export type TraceEntryKind =
  | "agent"
  | "tool"
  | "guard"
  | "state_event"
  | "retrieval"
  | "memory";

export interface TraceStep {
  kind: TraceEntryKind;
  name: string;
  summary?: string;
  detail?: Record<string, unknown>;
}

export interface AgentTrace {
  turn_id: string;
  mode: string;
  route: string;
  risk_level: string;
  agents: TraceStep[];
  tools: TraceStep[];
  guards: TraceStep[];
  state_event: TraceStep | null;
  memory_used: boolean;
  retrieval_used: boolean;
  safety_critic_used: boolean;
  conversation_history_used: boolean;
  conversation_history_count: number;
  conversation_seed_used: boolean;
  conversation_seed_count: number;
  research_metadata: Record<string, unknown>;
  research_trace?: ResearchTraceMetadata;
}

export interface ResearchTraceMetadata {
  interaction?: {
    intent: string | null;
    role_selection_source: string | null;
    context_role_ids: string[];
  };
  role: {
    candidate_roles?: string[];
    selected_roles: string[];
    silent_roles?: string[];
    primary_role: string | null;
    role_selection_mode: string | null;
    requested_role_ids: string[];
    cueing_style: string | null;
    allow_follow_up?: boolean;
    follow_up_reason?: string | null;
  };
  topic: {
    topic_id: string | null;
    topic_label: string | null;
    material_type: string | null;
    classified_topic: string | null;
  };
  boundary: {
    boundary_state: string;
    boundary_notes: string[];
  };
  control: {
    study_condition: string | null;
    study_session_id: string | null;
    elder_control_action: string | null;
  };
}

// A persisted trace returned by GET /api/traces/{turn_id} (#9).
export interface TraceRecord {
  turn_id: string;
  user_id: string;
  created_at: string;
  trace: AgentTrace;
}

// A compact row from GET /api/traces?user_id=&limit= (#9).
export interface TraceSummary {
  turn_id: string;
  user_id: string;
  created_at: string;
  route: string;
  risk_level: string;
  safety_critic_used: boolean;
  retrieval_used: boolean;
}
